### ISSUE TITLE
Telegram Notifications Module not sending notifications to group bot chat

### ISSUE THREAD

**Issue by jordanurbs** [freescout-helpdesk/freescout#2507]
Purchased a license and followed the directions precisely (even tried setting up a second bot), but despite no errors I am not receiving any messages from the bot on Telegram. No clues in the app logs, either.

I am receiving emails in the inboxes no problem... just no Telegram

**Comment by freescout-helpdesk** [comment]
Could you please attach screenshot of Telegram settings page.

**Comment by jordanurbs** [comment]
<img width="1259" alt="Screen Shot 2023-01-04 at 9 49 26 AM" src="https://user-images.githubusercontent.com/44348964/210618413-25666718-4117-4170-b0e7-2bd95f89bc4f.png">

**Comment by freescout-helpdesk** [comment]
Settings look fine. If you want you can provide access to your server (email at support@freescout.net) and we'll check it. Otherwise you may need to debug integration yourself.

**Comment by jordanurbs** [comment]
> Settings look fine. If you want you can provide access to your server (email at [support@freescout.net](mailto:support@freescout.net)) and we'll check it. Otherwise you may need to debug integration yourself.

Could you please point me in the direction of debug integration? I'm running Freescout on Cloudron and am not sure providing access to the server would be so simple.

**Comment by freescout-helpdesk** [comment]
You can also check if there are any failed background jobs on Manage > Status page.

The code sending notifications is located in /Modules/Telegram/Providers/TelegramServiceProvider.php (line 279)

**Comment by jordanurbs** [comment]
There are no failed background jobs. Everything else is green.

Am I looking for something in particular in this code?

`try {
                \TelegramBot::sendMessage([
                    'chat_id' => $chat_id,
                    'parse_mode' => 'HTML',
                    'text' => $message,
                    'disable_web_page_preview' => true
                ]);
            } catch (\Exception $e) {
                \Helper::log('telegram', 'API error: '.$e->getMessage());
            }`

**Comment by freescout-helpdesk** [comment]
Yes. It looks like you don't have enough qualification to debug the issue. Try to find someone who can do it for you.

**Comment by nebulade** [comment]
I'm coming here from https://forum.cloudron.io/topic/8381/telegram-notifications-module-compatibility ...

Is there a way to get a temporary or dev license key for this module? Then we can debug this from a Cloudron packaging perspective.

**Comment by freescout-helpdesk** [comment]
Sure. Just message at support@freescout.net

**Comment by nebulade** [comment]
Thanks, but I went ahead and bought a module license now to debug this. I also managed to setup the token and freescout does fetch the available channels just fine. Only no notifications were sent, however after restarting the app (in this case through Cloudron dashboard) it started working and it seems it requires an `artisan schedule:run` restart to pick up the settings or start triggering the notification. After that it works fine.

**Comment by nebulade** [comment]
That is good to know. In the Cloudron package case we run it via supervisor and the config is https://git.cloudron.io/cloudron/freescout-app/-/blob/master/supervisor/artisan-queue.conf so maybe freescout settings simply has no clue on how to restart this properly in such an environment? If that is the case, we have to see how to fixup our package then, as this might cause other side-effects as well.

**Comment by freescout-helpdesk** [comment]
https://github.com/freescout-helpdesk/freescout/blob/8a4649ceae62c8f8eff4842f10df0802953d3c4e/app/Console/Kernel.php#L152
This line starts `queue:work` as a sub-process. It sets a special mutex in the cache, and when the cache is cleared this mutex is removed and `queue:work` stops and on the next cron launch another `queue:work` is executed.

**Comment by nebulade** [comment]
Not sure I really understand the construct here, however also the reversed is the case. Clearing the bot token field in the freescout settings, does update the env file, but the app still sends out notification until that process is restarted. Unfortunately I am not seeing any events or logs about some `queue:work` or so being restarted on settings save.

**Comment by freescout-helpdesk** [comment]
Also in cron.sh currently there are just few commands: https://git.cloudron.io/cloudron/freescout-app/-/blob/master/cron.sh
Some commands from https://github.com/freescout-helpdesk/freescout/blob/dist/app/Console/Kernel.php are missing and also some modules should be able to run their commands using the hook in `schedule:run`: https://github.com/freescout-helpdesk/freescout/blob/dist/app/Console/Kernel.php#L115

So we can add some parameter for `artisan schedule:run` which would allow just to execute all the commands in https://github.com/freescout-helpdesk/freescout/blob/dist/app/Console/Kernel.php and skip executing 'queue:work'. Will this work?

**Comment by freescout-helpdesk** [comment]
We've made some updates - when settings are saved and the cache is cleared the 'queue:work' which is currently running will stop by itself. And Cloudron supervisor should restart it. It should allow to fix Issue 1).

**Comment by nebulade** [comment]
Actually before we try to work around in freescout code, maybe our packaging can just be fixed up. Possibly the usage of this long running `schedule:run` through supervisor was simply done wrong by us? Ideally we can stick as close as possible to what your preferred way here is, to avoid future issues. So is it better to only have the `cron.php` with other commands and besides that no extra long-running process?

**Comment by freescout-helpdesk** [comment]
Standard approach is just to have this command ran by cron:
```
* * * * * php /var/www/html/artisan schedule:run >> /dev/null 2>&1
```
It takes care of everything by itself - runs all the necessary commands every minute and maintains long-running `queue:work` process.

**Comment by nebulade** [comment]
Ok so just to be on the same page, we require a `queue:work` process run in a long lived fashion, which if for some reason stops, should be restarted to run again. This restart is ensured normally by the cron job running `schedule:run`. The `watch:queue` might get killed (or stops automatically via the mutex) triggered by some things like the settings (in the env file) changes?

Now since on Cloudron we run the cron job in a separate docker container, we don't want that to be long-lived so I wonder if we ensure the `watch:queue` process to be always running through supervisor in the main app container, then we don't need a cron job at all?

If this is all correct, then the question remains, if that `watch:queue` restart works as expected in such a case. I'm trying to put more debug logs there, to see if the restart works, but I am not really sure where to put those. Can you give some guidance here?

**Comment by freescout-helpdesk** [comment]
> The watch:queue might get killed (or stops automatically via the mutex) triggered by some things like the settings (in the env file) changes?

Yes. You probably mean `queue:work` (not `watch:queue`).

There is no way to avoid running `artisan schedule:run` by cron as it executes number of commands which should be executed every minute. The `queue:work` does not replace it, it does it's own work.

Improved way of restarting the `queue:work` has already been released in FreeScout v1.8.48. You can just update to the latest version and make sure Telegram settings are applied.

**Comment by nebulade** [comment]
I am updating the package right now and so you are suggesting to replace the individual commands in our cron.sh with `artisan schedule:run` while also having a supervisor daemon running `artisan schedule:run` which will act as the long-lived process? I hope there is no case then where the cron takes over the long-lived process via `artisan schedule:run` at some point. Maybe there is a flag to run it from cron to never daemonize?

**Comment by freescout-helpdesk** [comment]
> Maybe there is a flag to run it from cron to never daemonize?

Not yet. We'll try to add it.

**Comment by freescout-helpdesk** [comment]
Implemented in the `master` branch. If `--no-interaction` flag is set the `queue:work` daemon will not be executed:
```
php artisan schedule:run --no-interaction
```

**Comment by nebulade** [comment]
Great! Then we have to remove the `--no-interaction` from the supervisor config, hopefully that does not trigger any required user/cli interaction though.

**Comment by freescout-helpdesk** [comment]
`schedule:run` does not trigger any interaction.

**Comment by nebulade** [comment]
seems like our license key could only be applied once to test this initial setup behaviour. If possible could you send us a dev one to support@cloudron.io then I will go ahead and test the full flow and if working, we will wait for the next freescout release to finally support the new `--no-interaction` behaviour properly. 

Thanks already for the great support in getting this sorted out!

**Comment by nebulade** [comment]
Great I can confirm that it works now. We will update the package once FreeScout 1.8.49 is officially released. I guess this can be closed then. Thanks everyone for getting this sorted out.

**Comment by freescout-helpdesk** [comment]
Update has been released in FreeScout v1.8.49

**Comment by jordanurbs** [comment]
Just want to confirm that everything is now working as expected. Thanks everyone.
